// @grammyjs/commands/out/deps.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.P = exports.match = exports.Context = exports.Composer = exports.Bot = exports.Api = void 0;
// TODO: Replace with official deno module, once it arrives (https://github.com/gvergnaud/ts-pattern/pull/108)
var grammy_1 = require("grammy");
Object.defineProperty(exports, "Api", { enumerable: true, get: function () { return grammy_1.Api; } });
Object.defineProperty(exports, "Bot", { enumerable: true, get: function () { return grammy_1.Bot; } });
Object.defineProperty(exports, "Composer", { enumerable: true, get: function () { return grammy_1.Composer; } });
Object.defineProperty(exports, "Context", { enumerable: true, get: function () { return grammy_1.Context; } });
var ts_pattern_1 = require("ts-pattern");
Object.defineProperty(exports, "match", { enumerable: true, get: function () { return ts_pattern_1.match; } });
Object.defineProperty(exports, "P", { enumerable: true, get: function () { return ts_pattern_1.P; } });
